src/chain/optimism.rs

```diff
@@ -0,0 +1,286 @@
+    }
+}
+
+/// Convert [Transaction] to [OptimismFields]
+pub(crate) fn get_optimism_fields(
+    tx: &Transaction,
+) -> Result<OptimismFields, OptimismTransactionParsingError> {
+    let envelope_buf = {
+        let tx_type = tx.inner.transaction_type.unwrap_or_default();
+        let op_tx_type = OpTxType::try_from(tx_type)
+            .map_err(|_err| OptimismTransactionParsingError::UnexpectedType(tx_type))?;
+        let inner = tx.inner.clone();
+        let tx_envelope = match op_tx_type {
+            OpTxType::Legacy => Signed::<TxLegacy>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Eip2930 => Signed::<TxEip2930>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Eip1559 => Signed::<TxEip1559>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Eip4844 => Signed::<TxEip4844>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Deposit => {
+                let tx_deposit = TxDeposit {
+                    source_hash: tx
+                        .source_hash
+                        .ok_or(OptimismTransactionParsingError::MissingSourceHash)?,
+                    from: tx.inner.from,
+                    to: tx.inner.to.into(),
+                    mint: tx.mint,
+                    value: tx.inner.value,
+                    gas_limit: tx.inner.gas,
+                    is_system_transaction: tx.is_system_tx.unwrap_or_default(),
+                    input: tx.inner.input.clone(),
+                };
+                Ok(OpTxEnvelope::from(tx_deposit))
+            }
+        }
+        .map_err(|err| OptimismTransactionParsingError::ConversionError(err.to_string()))?;
+
+        let mut envelope_buf = Vec::<u8>::new();
+        tx_envelope.encode_2718(&mut envelope_buf);
+        Bytes::from(envelope_buf)
+    };
+
+    Ok(OptimismFields {
+        source_hash: tx.source_hash,
+        mint: tx.mint,
+        is_system_transaction: tx.is_system_tx,
+        enveloped_tx: Some(envelope_buf),
+    })
+}
+
+impl PevmChain for PevmOptimism {
+    type Transaction = Transaction;
+    type BlockSpecError = OptimismBlockSpecError;
+    type TransactionParsingError = OptimismTransactionParsingError;
+
+    fn id(&self) -> u64 {
+        self.id
+    }
+
+    fn build_tx_from_alloy_tx(&self, tx: alloy_rpc_types::Transaction) -> Self::Transaction {
+        Transaction {
+            inner: tx,
+            ..Transaction::default()
+        }
+    }
+
+    fn get_block_spec(&self, header: &Header) -> Result<SpecId, Self::BlockSpecError> {
+        if header.timestamp >= 1720627201 {
+            Ok(SpecId::FJORD)
+        } else if header.timestamp >= 1710374401 {
+            Ok(SpecId::ECOTONE)
+        } else if header.timestamp >= 1704992401 {
+            Ok(SpecId::CANYON)
+        } else if header.number >= 105235063 {
+            Ok(SpecId::REGOLITH)
+        } else {
+            // TODO: revm does not support when L1Block is not available
```

Comment:
There's no default value like `BEDROCK`? Maybe a link to an `revm` issue or code would help clarify this.

Reply:
`SpecId::BEDROCK` exists but `revm` is not handling a Bedrock block correctly. Therefore, I throw an error here to stop execution immediately.

Let me verify again the `revm` issue when handling a Bedrock block.

Reply:
That sounds like something we should definitely have some doc for.

Reply:
Sorry for the confusion.
- On Optimism "Testnet" (not sure if it is Optimism Sepolia or whatever): Regolith is activated after Bedrock.
- On Optimism Mainnet, Regolith and Bedrock are activated at the same time (block number = 105235063).

As a result, this fn should never return `SpecId::BEDROCK`.

------

Based on `reth` repo, the implementation should be:

```rust
    fn get_block_spec(&self, header: &Header) -> Result<SpecId, Self::BlockSpecError> {
        // TODO: The implementation below is only true for Optimism Mainnet.
        // When supporting other networks (e.g. Optimism Sepolia), remember to adjust the code here.
        if header.timestamp >= 1720627201 {
            Ok(SpecId::FJORD)
        } else if header.timestamp >= 1710374401 {
            Ok(SpecId::ECOTONE)
        } else if header.timestamp >= 1704992401 {
            Ok(SpecId::CANYON)
        } else if header.number >= 105235063 {
            Ok(SpecId::REGOLITH)
        } else if header.number >= 3950000 {
            Ok(SpecId::BERLIN)
        } else {
            Ok(SpecId::MUIR_GLACIER)
        }
    }
```

However, `revm` does not seem to support pre-Bedrock blocks (i.e. when block number < 105235063).

PS: This observation is the same as my previous observation a long time ago. Basically, I was just redoing whatever I have done previously.

Reply:
Added some comments to explain.